### ISSUE TITLE
Doc: styles and stylesheet handling

### ISSUE THREAD

**Issue by poire-z** [koreader/crengine#167]
Some notes I wrote while looking at src/lvstsheet.cpp:

## What is what:

**Declaration**: `{ color: red; margin: 10px }`
**Selector**: `div.intro` or `div.intro[attr=val]`
**SelectorRule**: `.intro`, `.intro[attr=val]`, `#TOP`
**ElementName/ElementNameID**: internal crengine id identifying each element (eg: P=12, DIV=13, SPAN=14...)
**NodeID**: internal id identifying a node (2 different DIV nodes have the same ElementName (13))

## LVCssDeclaration 
```
class LVCssDeclaration {
    int * _data; // packed (3 int per property) representation of parsed string
            in the order seen, including duplicates
```

`LVCssDeclaration.parse("{ color: red; margin: 10px }")` pack string into _data

`LVCssDeclaration.apply( css_style_rec_t )` => apply to cre node style object
     process packed _data in its order, simply transfer each property to the css_style_rec_t struct

With #166: _"!important" is now parsed here, and when apply(), flagged for each prop of css_style_rec_t
When apply(), check for that flag and do not transfer/override if flagged (unless also important)_

Note: src/lvrend.cpp : only usage of LVCssDeclaration outstide of lvstsheet.cpp, to parse node inlined style="color: blue; display: inline;":
https://github.com/koreader/crengine/blob/55f9f806f7e468ccd60254b608c9c95e0b2886e9/crengine/src/lvrend.cpp#L2704-L2715

## LVCssSelectorRuleType
enum cssrt_ancessor, cssrt_id, cssrt_class...

## LVCssSelectorRule
```
    class LVCssSelectorRule {          // no reference to current element id (P, H1)
        LVCssSelectorRuleType _type;   //   for H1>P[foo="bar"] :
        lUInt16 _id;                   //     ElementNameID for H1 : ID of parent/ancestor (so NULL for simple rules)
        lUInt16 _attrid;               //     cre attributeId for "foo"
        LVCssSelectorRule * _next;     // (these are chained, if multiple rules: all should match)
        lString16 _value;              //     cre attributeValueId for "bar"

```
`LVCssSelectorRuleType.check(ldomNode)`  // check if match, by looking at node attributes, class, parent id, class, #id

## LVCssSelector
```
    class LVCssSelector {
        lUInt16 _id;                    // ElementNameID (P=12), or 0 for ".className" or "*" selectors
        LVCssDeclRef _decl;
        int _specificity;               // not implemented, always 0
        LVCssSelector * _next;          // chained (by LVStyleSheet.parse()), when:
                                        //   el1, el2.cls3 { decl }: will point to same _decl
        LVCssSelectorRule * _rules;     // chained, when multiple: element1 element2.class3 { } (without ',')
                                        // chained => check(node) should match ALL
                                        // can be NULL => check(node) match simply when _id matches ElementNameID
```

`LVCssSelector.setDeclaration( LVCssDeclRef )`
  { _decl = decl; }

`LVCssSelector.parse( str, lxmlDocBase );`
  parse str, and return after first selector found (till "," or "{" happens)
  set _id to the matched ElementName (0 if ".className" or "*")
  create 0 or many LVCssSelectorRule and add them to the _rules chain

`LVCssSelector.check( ldomNode );`
  check _id (if !=0) matches node ElementName
  call LVCssSelectorRuleType.check(node) on each rule

`LVCssSelector.apply( ldomNode, css_style_rec_t)`
  { if (check( node )) _decl->apply(style); }

## LVStyleSheet
```
    lxmlDocBase * _doc;     // document to retrieve ID values from
    LVPtrVector <LVCssSelector> _selectors;               // list of LVCssSelector, indexed by elementName ID:
                                                          //  so: 0 .. max nb of ElementNames
                                                          // 0 contains all the Selector without an element name (.cls, #top)
                                                          // 12 would contain all the Selectors like P P.cls P[foo=bar]..
    LVPtrVector <LVPtrVector <LVCssSelector> > _stack;    // stack of list of LVCssSelector, for push/pop

```
`LVStyleSheet.push() { _stack.add( dup() ); }` // save (copy, current is kept) current state of stylesheet
`LVStyleSheet.pop() { v = _stack.pop(); set( *v ); delete v; }` // restore previously saved state

`LVStyleSheet.parse( const char * str );` /// parse stylesheet, compile and add found rules to sheet
    parse whole str full stylesheet
    parse selectors:
        create one LVCssSelector for each el1, el2.cls3
    parse declaration:
        create one LVCssDeclaration, set it to all created selectors
    for each LVCssSelector:
      grow _selectors[LVCssSelector._id] if needed
      insert/append to _selectors[LVCssSelector._id], most specificity-ic (not implemented) first

`LVStyleSheet.apply( ldomNode, css_style_rec_t );` /// apply stylesheet to node style
    call LVCssSelector.apply for each LVCssSelector, in that order:
      first, those from _selectors[_nodeID] (_selectors[12] for P): so, elementNames first
      then, those from _selectors[0] (classNames and *): so, elementNameLess last
    taking into account (not implemented) specificity that could switch that order
https://github.com/koreader/crengine/blob/55f9f806f7e468ccd60254b608c9c95e0b2886e9/crengine/src/lvstsheet.cpp#L2340-L2370

Still investigating this stack/push/pop stuff.

## Other stuff
Applying stylesheet to a node is just a small part of its style initialization. Inheritance and other stuff are done in:
https://github.com/koreader/crengine/blob/55f9f806f7e468ccd60254b608c9c95e0b2886e9/crengine/src/lvrend.cpp#L2678-L2861

**Comment by Frenzie** [comment]
Very nice. :-)

Definitely a pity about specificity though. You'd could easily end up with a weird situation like this:

```css
#some_id.some_class.some_other_class.some_other_class_still table ul p {border: 1px solid black}
p {border: 20 px dotted green}
```

Now I'd argue that a well-written stylesheet naturally progresses from less specific toward more specific. That is, you start with basic element styles before moving on to the "real" styles, so if we naïvely assume relatively well-structured stylesheets the damage should be somewhat limited, but still. :-)

**Comment by poire-z** [comment]
Well, there is already some code to process selectors by ascending order of specificity. It's just that `LVCssSelector._specificity` is currently always 0.
I guess it shouldn't be too hard to increment that _specificty according to elementName/elementName-less and the various LVCssSelectorRule as they are added to the LVCssSelector, following https://www.w3.org/TR/CSS2/cascade.html#specificity

**Comment by Frenzie** [comment]
Huh, I had no idea there was so much stuff hiding out just on the verge of greatness. ;-)

**Comment by poire-z** [comment]
Possibly https://github.com/koreader/crengine/commit/d795b6b486c5fb6d31c08f9f08a00387e8ef2832.

There is some check and warning about DRM in crengine code:
https://github.com/koreader/crengine/blob/89b0650e425013da81bc73c6d7676599d8b13c62/crengine/src/epubfmt.cpp#L717-L728
https://github.com/koreader/crengine/blob/89b0650e425013da81bc73c6d7676599d8b13c62/crengine/src/epubfmt.cpp#L391-L404

**Comment by Frenzie** [comment]
No, that's not it. It's actually something post-v2018.11. Here's calibre + inspector + crengine misrendering in the October 2018 release. I'm guessing it's #241.

![screenshot_2018-12-03_19-59-02](https://user-images.githubusercontent.com/202757/49395127-f1fcbd00-f735-11e8-821d-6bc320211a0f.png)

Then in v2018.12 it looks completely different (and correct):
![screenshot_2018-12-03_20-02-25](https://user-images.githubusercontent.com/202757/49395298-633c7000-f736-11e8-849a-0e40457bfc19.png)

**Comment by Frenzie** [comment]
No, actually that's not it either. Darn, I'm really going to have to revert all of those if I want to know, aren't I. :-P

(Unless you know based on the screen?)

![screenshot_2018-12-03_20-09-25](https://user-images.githubusercontent.com/202757/49395638-5a986980-f737-11e8-9523-d19a7d3aacb4.png)

![screenshot_2018-12-03_20-09-11](https://user-images.githubusercontent.com/202757/49395637-5a986980-f737-11e8-8aed-6055347059e9.png)

**Comment by poire-z** [comment]
I too want to know :)
But nothing obvious from your screenshots ... the HTMLs look the same.

Are there some other styles like `_idGenCharOverride-12` `_idGenCharOverride-123` or `CharOverride-1`?
How is the vertical spacing for the following lines after the BIG words in the old version (just to see if it's related to some line-height / strut fixes).
No style tweak in play?

**Comment by Frenzie** [comment]
No, I tested with vanilla AppImages. Plus my actual reader is fairly vanilla too. ;-)

```css
p.ParaOverride-1 {
	text-align:center;
}
p.ParaOverride-2 {
	margin-left:4px;
}
p.ParaOverride-3 {
	text-indent:36px;
}
p.ParaOverride-4 {
	-epub-hyphens:none;
	-webkit-hyphens:none;
}
span.CharOverride-1 {
	font-family:Gothup;
	font-size:4em;
	font-style:normal;
	font-weight:normal;
}
span.CharOverride-2 {
	font-family:Palatino;
}
span._idGenCharOverride-1 {
	font-family:Palatino;
}
div._idGenObjectStyleOverride-1 {
	border-width:0px;
}
```

**Comment by poire-z** [comment]
So you found it :)
```css
span.CharOverride-1 {
	font-family:Gothup;
	font-size:4em;   /* yeah, 4em in the morning */
	font-style:normal;
	font-weight:normal;
}
```
was probably matched and used by `<span class="Empasis _idGenCharOverride-1">` because of the classname check bug fixed in #241.
